### PR TITLE
Add [Entity] and [Index] in order to automatically ensure indexes of all entities

### PR DESCRIPTION
**Pull request by pingvinen**
Add [Entity] and [Index] in order to automatically ensure indexes of all entities

We want an easy way to declare and ensure indexes on MongoDB collections. The path chosen for this involves marking your entities with [Entity] and the properties you wish indexed marked with [Index].

Compound indexes are also supported - even across inheritance levels and nested documents. Documentation will follow in the docs repository.

An example:

```csharp
namespace My.Entities
{
	[Entity]
	public class AwesomeEntity
	{
		[Index]
		public string MustBeFast { get; set; }

		public SubDocument Nested { get; set; }
	}

	public class SubDocument
	{
		[Index("with_a_name")]
		public int Important { get; set; }
	}
}

namespace My.Main
{
	public class SetItAllUpAndStuff
	{
		public void Start()
		{
			var client = new MongoClient(connectionString);
			new MongoManager().EnsureIndexes(
				client.GetDatabase("SoAwesome"),
				typeof(AwesomeEntity).GetTypeInfo().Assembly,
				typeof(AwesomeEntity).GetTypeInfo().Namespace
			);
		}
	}
}
```

### CODE REVIEW

**Comment by pingvinen** [comment]
@nover We should also find a way to run the functional tests during the AppVeyor build. There is a docker-compose file, so we just need to start that.

**Comment by nover** [comment]
@pingvinen yeah we need something there - unfortunately it does not look like AppVeyor have capability for starting docker containers and linking them during the test phase.

File: src/Internal/IndexDefinition.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        /// Generates an IndexKeysDefinition instance for this index
+        /// </summary>
+        /// <returns>BsonDocumentIndexKeysDefinition<DocumentType></returns>
+        public virtual object GetKeySpec()
+        {
+            var indexDefinitionsType = typeof(BsonDocumentIndexKeysDefinition<>).MakeGenericType(DocumentType);
+            return Activator.CreateInstance(indexDefinitionsType, GetKeySpecBsonDocument());
+        }
+
+        public virtual CreateIndexOptions GetOptions()
```
Comment (nover) [inline src/Internal/IndexDefinition.cs]: Perhaps some XML doc on the public members, or is the method name self-explanatory enough?

Comment (pingvinen) [inline src/Internal/IndexDefinition.cs]: Could be. Remember, though, that it is in the Ìnternal` namespace, and thus should not be used outside the repository.

**Comment by pingvinen** [comment]
There is actually a bug in how the code generates nested indexes.

**Comment by pingvinen** [comment]
Fixed the bug with not including a path to nested field in the keys.

**Pull request merged**